whylog/constraints/__init__.py

```diff
@@ -2,42 +2,177 @@
 
 import six
 
+from whylog.constraints.const import ConstraintType
+from whylog.constraints.exceptions import ConstraintVerificationError, ConstructorParamsError
+from whylog.teacher.user_intent import UserConstraintIntent
+
 
 @six.add_metaclass(ABCMeta)
 class AbstractConstraint(object):
+    """
+    :param MIN_GROUPS_COUNT: minimal groups count needed to create constraint
+    :param MAX_GROUPS_COUNT: maximal groups count needed to create constraint
+    """
+
+    TYPE = 'undefined'
+    MIN_GROUPS_COUNT = None
+    MAX_GROUPS_COUNT = None
+
     @abstractmethod
-    def save(self, rulebase_rule):
-        pass
+    def __init__(self, param_dict, groups):
+        """
+        Creates constraint object.
+
+        For Teacher and Front use while creating user rule.
+
+        :param param_dict: dict of additional params of constraint
+        :param groups: list of tuples (line_id, group_no),
+                       where line_id and group_no is inner numeration between Front and Teacher
+        """
+        param_dict_keys = param_dict.keys()
+        correct_param_names = self.get_param_names()
+        if not param_dict_keys == correct_param_names:
```

Comment:
Why we can't implement mechanism for optional params? In your implementation all params are required.